sycl/doc/extensions/proposed/sycl_ext_oneapi_non_uniform_groups.asciidoc

```diff
@@ -0,0 +1,1161 @@
+incompatible ways before it is finalized.  *Shipping software products should
+not rely on APIs defined in this specification.*
+
+
+== Overview
+
+Many modern hardware architectures support flexible sub-divisions of
+work-groups and sub-groups to support fine-grained work scheduling. A common
+use-case for such flexibility is communication and synchronization between
+work-items in divergent control flow.
+
+This proposal introduces new classes to represent sub-divisions of SYCL's
+built-in group types, traits for detecting these classes, and free functions
+for creating new instances of these classes.
+
+These new classes can be used as arguments to group functions, group algorithms,
+and custom functions to convey exactly which work-items an operation is expecting,
+simplifying interfaces and greatly reducing the amount of delicate wording
+needed in function documentation.
+
+NOTE: The first version of this extension only supports partitioning of
+sub-groups. It is expected that in the future, these functions will be expanded
+to also allow partitioning of root-groups, work-groups and user-constructed
+groups.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_NON_UNIFORM_GROUPS` to one of the values defined in the
+table below.  Applications can test for the existence of this macro to
+determine if the implementation supports this feature, or applications can test
+the macro's value to determine which of the extension's features the
+implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== Control Flow
+
+The SYCL specification defines
+link:https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#control-flow[control
+flow] as below:
+
+> When all work-items in a group are executing the same sequence of statements,
+> they are said to be executing under _converged_ control flow. Control flow
+> _diverges_ when different work-items in a group execute a different sequence of
+> statements, typically as a result of evaluating conditions differently (e.g.
+> in selection statements or loops).
+
+This extension introduces some new terminology to describe other kinds of
+control flow, to simplify the description of the behavior for new group types.
+
+A work-item is said to have _active_ control flow for a given point of
+execution if it follows the sequence of statements currently being executed.
+
+Two or more work-items are considered to have _simultaneous_ execution at a given
+point if they execute together. The amount of and granularity of simultaneous
+execution is implementation-defined.
+
+A _tangle_ is a collection of work-items from the same sub-group executing
```

Comment:
Thank you for reviewing so thoroughly and asking all these questions.

When we wrote this initially, we decided a tangle was a subset of a sub-group.  But re-reading the definition of "tangled instruction" [from SPIR-V](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#TangledInstruction), I now see that they include "Group and subgroup instructions"...  I still have some reservations about exposing a group type that represents all members of a work-group in the same control flow (because I'm not sure it's implementable), but if we keep the name "tangle" and deviate from SPIR-V it is bound to confuse somebody.

Since this is an experimental extension, would it be acceptable to you to add a template parameter for now and say that the group supports work-groups?  I suspect we'll want to revisit this (and maybe change the name as a result), but we should probably make that decision based on real implementation experience.